### PR TITLE
Preview window

### PR DESCRIPTION
**Pull request by K1llMan**
Preview window with frame selector and test button for single frame.
image

### CODE REVIEW

**Comment by henryruhs** [comment]
Can you resolve conflicts... can roop.ui be made fully functional instead of a class?

File: roop/core.py [hunk 1]
```diff
@@ -16,13 +16,13 @@
...
 import threading
-from threadpoolctl import threadpool_limits
 from PIL import Image, ImageTk
 
 import roop.globals
-from roop.swapper import process_video, process_img
+from roop.swapper import process_video, process_img, process_faces
 from roop.utils import is_img, detect_fps, set_fps, create_video, add_audio, extract_frames, rreplace
 from roop.analyser import get_face_single
+import roop.ui as ui
```
Comment (henryruhs) [inline roop/core.py:25]: Thanks a lot for refactoring it to a file

File: roop/core.py [hunk 2]
```diff
@@ -129,6 +143,26 @@ def preview_video(video_path):
...
+        # Preview
+        preview.update(frame)
+        amount_of_frames = cap.get(cv2.CAP_PROP_FRAME_COUNT)
+        
+        def update_slider(frame_number):
+            preview.update(get_video_frame(video_path, frame_number))
+
+        preview.init_slider(amount_of_frames, update_slider)
+
+        def test_handler():
```
Comment (henryruhs) [inline roop/core.py:155]: Please no nested def... move it out - can you rename it preview_handler() aswell?

**Comment by K1llMan** [comment]
> can roop.ui be made fully functional instead of a class?

It's hard to manage nested elements and frames to make correct view. I think logic and view must be separated. If we can use model to pass data from functions and render views by model then it should be easier to manage parameters and its visual presentation.

**Comment by henryruhs** [comment]
Okay then, but could you still be so awesome and move everything to ui.py and proably turn it to functional code? This would be a great help for me.

**Comment by K1llMan** [comment]
> Okay then, but could you still be so awesome and move everything to ui.py and turn your class it to functional code?

Ok. Could you please merge changes before it? A lot of time needed to merge change from there and back again and again. I'll make ui refac on another PR.

**Comment by henryruhs** [comment]
@K1llMan No worries I will take care that it does not break another time - I just cannot merge unfinished PRs as this will add more work on my table to get it to a state that can land in main. That being said, I know it is annoying but please go the extra mile.

**Comment by henryruhs** [comment]
Pipeline fails with error:

```
Status: swapping in progress...
    start()
  File "/home/runner/work/roop/roop/roop/core.py", line 252, in start
    process_video(args['source_img'], args["frame_paths"], preview.update)
NameError: name 'preview' is not defined
```

**Comment by K1llMan** [comment]
Should be fixed.

**Comment by K1llMan** [comment]
All moved to ui and be functional.

**Comment by henryruhs** [comment]
@s0md3v did not have a chance to test it but looks ready to be merged as of the code changes... having everything in `ui.py` is a big step forward

@K1llMan Good work 👍

**Comment by s0md3v** [comment]
It works. The test button should let the user know the test was successful but that can be done later.

**Comment by henryruhs** [comment]
@K1llMan I'm very thankful for this addition, again great work!

**Pull request merged**